src/main/java/org/jetbrains/bsp/bazel/server/BepServer.java

```diff
@@ -266,132 +234,98 @@ private void processCompletionEvent(BuildEventStreamProtos.BuildEvent event) thr
+        diagnosticsProtosLocations.remove(target.getUri());
+      }
 
-      System.out.println("Found diagnostics file in " + log.getUri());
-      if (hasDiagnosticsOutput) diagnosticsProtosLocations.remove(target.getUri());
-      getDiagnostics(filesToDiagnostics, target, log.getUri().substring(7));
+      diagnosticsDispatcher.collectDiagnostics(
+          filesToDiagnostics, target, log.getUri().substring(7));
     }
 
     if (filesToDiagnostics.isEmpty() && hasDiagnosticsOutput) {
-      getDiagnostics(filesToDiagnostics, target, diagnosticsProtosLocations.get(target.getUri()));
+      diagnosticsDispatcher.collectDiagnostics(
+          filesToDiagnostics, target, diagnosticsProtosLocations.get(target.getUri()));
       diagnosticsProtosLocations.remove(target.getUri());
     }
 
-    emitDiagnostics(filesToDiagnostics, target);
+    diagnosticsDispatcher.emitDiagnostics(filesToDiagnostics, target);
   }
 
-  public void emitDiagnostics(
-      Map<Uri, List<PublishDiagnosticsParams>> filesToDiagnostics, BuildTargetIdentifier target) {
-    for (SourceItem source : bspServer.getCachedBuildTargetSources(target)) {
-      Uri sourceUri = Uri.fromFileUri(source.getUri());
-      if (!filesToDiagnostics.containsKey(sourceUri)) {
-        filesToDiagnostics.put(
-            sourceUri,
-            Lists.newArrayList(
-                new PublishDiagnosticsParams(
-                    new TextDocumentIdentifier(sourceUri.toString()),
-                    target,
-                    new ArrayList<>(),
-                    true)));
-      }
-      if (bspClient != null) {
-        for (List<PublishDiagnosticsParams> values : filesToDiagnostics.values()) {
-          for (PublishDiagnosticsParams param : values) {
-            bspClient.onBuildPublishDiagnostics(param);
-          }
-        }
-      }
-    }
-  }
+  private HashMap<String, List<Diagnostic>> parseStdErrDiagnostics(
+      BuildEventStreamProtos.Progress progress) {
+    HashMap<String, List<Diagnostic>> fileDiagnostics = new HashMap<>();
 
-  public void getDiagnostics(
-      Map<Uri, List<PublishDiagnosticsParams>> filesToDiagnostics,
-      BuildTargetIdentifier target,
-      String diagnosticsLocation)
-      throws IOException {
-    Diagnostics.TargetDiagnostics targetDiagnostics =
-        Diagnostics.TargetDiagnostics.parseFrom(Files.readAllBytes(Paths.get(diagnosticsLocation)));
-    for (Diagnostics.FileDiagnostics fileDiagnostics : targetDiagnostics.getDiagnosticsList()) {
-      System.out.println("Inserting diagnostics for path: " + fileDiagnostics.getPath());
-      filesToDiagnostics.put(
-          Uri.fromExecOrWorkspacePath(
-              fileDiagnostics.getPath(),
-              bspServer.getBazelData().getExecRoot(),
-              bspServer.getBazelData().getWorkspaceRoot()),
-          convert(target, fileDiagnostics));
-    }
-  }
+    Arrays.stream(progress.getStderr().split("\n"))
+        .filter(
+            error ->
```

Comment:
I think that this predicate can be extracted too

Reply:
same as above, not sure about the names